dashboard/app/assets/javascripts/levels/contract_match.js.jsx

```diff
@@ -0,0 +1,257 @@
+
+  /**
+   * @param fn function to be curried with arguments
+   * @params [...] rest of args
+   * @returns {Function} new function with the given arguments pre-set as the
+   *                     defaults (curried)
+   */
+  var curry = function(fn) {
+    var args = Array.prototype.slice.call(arguments, 1);
+
+    return function () {
+      return fn.apply(this, args.concat(
+        Array.prototype.slice.call(arguments, 0)));
+    };
+  };
+
+  /**
+   * Enum of block types. Used for block and domain/range coloring
+   * @enum {string}
+   */
+  var blockValueType = {
+    NONE: 'None',
+    NUMBER: 'Number',
+    STRING: 'String',
+    IMAGE: 'Image',
+    BOOLEAN: 'Boolean'
+  };
+
+  var typesToColors = {};
+  typesToColors[blockValueType.NONE] = "#999999";
+  typesToColors[blockValueType.NUMBER] = "#00ccff";
+  typesToColors[blockValueType.STRING] = "#009999";
+  typesToColors[blockValueType.IMAGE] = "#9900cc";
+  typesToColors[blockValueType.BOOLEAN] = "#336600";
+
+  /**
+   * Component Structure:
+   *
+   * - ContractForm
+   *   - DomainsList
+   *     - TypeChooser
+   */
+  var ContractForm = React.createClass({
+    getName: function () {
+      return this.state.name;
+    },
+    getRangeType: function () {
+      return this.state.rangeType;
+    },
+    getDomainTypes: function () {
+      return this.state.domainTypes;
+    },
+    maxDomainID: 0,
+    getInitialState: function () {
+      /**
+       * @param data {{
+       *   name: string,
+       *   rangeType: BlockValueType,
+       *   domainTypes: {
+       *     type: BlockValueType,
+       *     key: string
+       *   },
+       * }}
+       */
+      return {
+        name: "",
+        rangeType: blockValueType.NONE,
+        domainTypes: [
+          {
+            key: 'domain' + (this.maxDomainID++),
+            type: blockValueType.NONE,
+            order: 0
+          }
+        ]
+      }
+    },
+    onNameChangeEvent: function (event) {
+      this.setState({
+        name: event.target.value
+      });
+    },
+    onRangeChange: function (newType) {
+      this.setState({
+        rangeType: newType
+      })
+    },
+    onDomainChange: function (domainKey, newType) {
+      this.setState({
+        domainTypes:
+          $.map(this.state.domainTypes, function (object) {
+            if (object.key == domainKey) {
+              object.type = newType;
+            }
+            return object;
+          })
+      });
+    },
+    onDomainAdd: function () {
+      var nextDomainID = this.maxDomainID++;
+      this.setState({
+        domainTypes:
+          this.state.domainTypes.concat({
+            key: 'domain' + nextDomainID,
+            type: blockValueType.NONE,
+            order: nextDomainID
+          })
+      });
+    },
+    onDomainRemove: function (domainKey) {
+      this.setState({
+        domainTypes:
+          $.grep(this.state.domainTypes, function (object) {
+            return object.key !== domainKey;
+          })
```

Comment:
This is perhaps a bigger discussion, but in places where there's both a Jquery and a lodash function available, I'd prefer to see us prioritize using lodash, and leave Jquery for things like easy http requests.